sky/spot/spot_state.py

```diff
@@ -419,9 +470,13 @@ def set_cancelling(job_id: int):
             (SpotStatus.CANCELLING.value, time.time(), job_id))
         if rows.rowcount > 0:
             logger.info('Cancelling the job...')
+    if callback_func is not None:
+        callback_func(-1, 'CANCELLING', '')
```

Comment:
Weirdly, when I am trying this PR, the `STARTING`, `RUNNING`, `RECOVERING`, `RECOVERED` works normally, while this `CANCELLING`/`CANCELLED` state was not sent to my slack. Do you have an idea why this happens?
Fyi, I did the following:
1. `sky spot launch test.yaml`
2. manually kill the cluster on the console
3. the states were correctly sent 
4. `sky spot cancel -a`: the cancelling/cancelled states were not correctly sent.

Reply:
I believe this might be because event_callback is non-blocking. Spot tasks could be killed first before the callback finishes execution.

Reply:
For `set_cancelling` outside of `SpotController`, I didn't pass in `event_callback_func`. 
https://github.com/skypilot-org/skypilot/blob/spot-event-callback/sky/spot/controller.py#L456

Reply:
Do we want to pass the event_callback_func to those as well? Otherwise, `sky spot cancel -a` will not send the message.

Reply:
I am thinking about a clean way to do this since we need to move `event_callback_func` outside of `SpotController`. Do you know if is there an easy way to get `task` from `task_id` and `job_id`? It seems to require `dag_yaml` or `self._dag`

Reply:
Ahh, as we did for the `_cleanup` function, we can retrieve the dag using the `day_yaml` as we did in `_cleanup` function https://github.com/skypilot-org/skypilot/blob/438f81599fd1728e5b8df6df7ca72e804bd6e1fd/sky/spot/controller.py#L430

However, we are not able to get some of the environment variable. We can probably print some log in the event callback saying which variables are not accessible, and we can wait for the user feedback to see if they want to have some of the variables in the CANCEL.

Reply:
@Michaelvll Thanks for all the detailed feedback! The suggested changes should all be incorporated